From 6ef47cf91985745b9aa861090008e1fd062c0b8e Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Fri, 14 Jun 2024 16:19:46 +0800
Subject: [PATCH 102/210] mtk: hostapd: find sta_info for legacy STA
 associating with non-first link when handling rx_from_unknown event

The legacy STA might associates with the AP MLD's non-first link, but
the 4-addr QoS NULL from STA is still sent to first link's hapd and
sta_info can not be found, leading to the 4-addr QoS NULL skb being
dropped.

This commit solves the problem by checking other link's hapd when the
sta_info is not found in handling rx_from_unknown event.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ieee802_11.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 70cd94083..eb0e137ad 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7257,6 +7257,21 @@ void ieee802_11_rx_from_unknown(struct hostapd_data *hapd, const u8 *src,
 	sta = ap_get_sta(hapd, src);
 
 #ifdef CONFIG_IEEE80211BE
+	if (!sta && hapd->conf->mld_ap) {
+		struct hostapd_data *h;
+
+		/* The data frame might be sent by a non-MLD STA via non-first
+		 * link, so we must also check other links.
+		 */
+		for_each_mld_link(h, hapd) {
+			if (h == hapd)
+				continue;
+
+			if (sta = ap_get_sta(h, src))
+				break;
+		}
+	}
+
 	if (sta && sta->mld_info.mld_sta)
 		sta = sta->mld_assoc_sta;
 #endif
-- 
2.45.2

